File: 2023/day1/Day1.scala
import scala.io.StdIn.readLine
import scala.annotation.tailrec

@main
def main =
  val res = solution
  println(s"Part 1: ${res._1}, Part 2: ${res._2}")

def solution =
  Iterator.continually(readLine())
    .takeWhile(s => s != null && !s.isEmpty())
    .map(s => (part1(s), part2(s)))
    .reduce((x, y) => (x._1 + y._1, x._2 + y._2))

def part1(s: String) =
  mkValue(s.filter(_.isDigit))

def part2(s: String) =
  mkValue(toDigits(s))

def mkValue(digits: String) =
  (digits.take(1) + digits.takeRight(1)).toIntOption.getOrElse(0)

def toDigits(s: String): String =
  toDigitsImpl("", s)

@tailrec
def toDigitsImpl(acc: String, s: String): String =
  if s.isEmpty() then acc
  else
    val accNew =
      if s.startsWith("one") then acc + "1"
      else if s.startsWith("two") then acc + "2"
      else if s.startsWith("three") then acc + "3"
      else if s.startsWith("four") then acc + "4"
      else if s.startsWith("five") then acc + "5"
      else if s.startsWith("six") then acc + "6"
      else if s.startsWith("seven") then acc + "7"
      else if s.startsWith("eight") then acc + "8"
      else if s.startsWith("nine") then acc + "9"
      else if s.head.isDigit then acc + s.head
      else acc

    toDigitsImpl(accNew, s.tail)

File: 2023/day1/Day1.hs
import Data.Char (isDigit)

main :: IO ()
main = interact solution

solution :: String -> String
solution s = show (part1 ls) ++ "\n" ++ show (part2 ls) ++ "\n"
  where
    ls = lines s
    part1 ls = sum $ map number ls
    part2 ls = part1 $ map replaceAll ls

number :: String -> Int
number s =
    let digits = filter isDigit s
        first = take 1 digits
        last = take 1 $ reverse digits
        num = if null digits then "0" else first ++ last
     in read num

replaceAll :: String -> String
replaceAll s = case replace s of
    (c : rest) -> c : replaceAll rest
    [] -> []

replace :: String -> String
replace s = case s of
    'o' : 'n' : 'e' : xs -> '1' : rest
    't' : 'w' : 'o' : xs -> '2' : rest
    't' : 'h' : 'r' : 'e' : xs -> '3' : rest
    'f' : 'o' : 'u' : 'r' : xs -> '4' : rest
    'f' : 'i' : 'v' : 'e' : xs -> '5' : rest
    's' : 'i' : 'x' : xs -> '6' : rest
    's' : 'e' : 'v' : 'e' : 'n' : xs -> '7' : rest
    'e' : 'i' : 'g' : 'h' : 't' : xs -> '8' : rest
    'n' : 'i' : 'n' : 'e' : xs -> '9' : rest
    _ -> s
  where
    rest = drop 1 s
